src/chaboche.jl

```diff
@@ -1,96 +1,167 @@
+    X2::Symm2{T} = fromvoigt(Symm2{T}, @view x[14:19])
+    return sigma, R, X1, X2
+end
+
+"""
+    integrate_material!(material::Chaboche{T}) where T <: Real
+
+Chaboche material with two backstresses. Both kinematic and isotropic hardening.
+"""
+function integrate_material!(material::Chaboche{T}) where T <: Real
     p = material.parameters
     v = material.variables
     dd = material.ddrivers
     d = material.drivers
     @unpack E, nu, R0, Kn, nn, C1, D1, C2, D2, Q, b = p
-    mu = E/(2.0*(1.0+nu))
-    lambda = E*nu/((1.0+nu)*(1.0-2.0*nu))
+    lambda, mu = lame(E, nu)
 
     @unpack strain, time = d
     dstrain = dd.strain
     dtime = dd.time
-    @unpack stress, X1, X2, plastic_strain, cumeq, R, jacobian = v
+    @unpack stress, X1, X2, plastic_strain, cumeq, R = v
 
-    jacobian = isotropic_elasticity_tensor(lambda, mu)
+    # elastic part
+    jacobian = isotropic_elasticity_tensor(lambda, mu)  # dσ/dε, i.e. ∂σij/∂εkl
+    stress += dcontract(jacobian, dstrain)  # add the elastic stress increment, get the elastic trial stress
 
-    stress += dcontract(jacobian, dstrain)
-    seff = stress - X1 - X2
-    seff_dev = dev(seff)
-    f = sqrt(1.5)*norm(seff_dev) - (R0 + R)
+    # resulting deviatoric plastic stress (accounting for backstresses Xm)
+    seff_dev = dev(stress - X1 - X2)
+    # von Mises yield function
+    f = sqrt(1.5)*norm(seff_dev) - (R0 + R)  # using elastic trial problem state
+    isplastic = (f > 0.0)
     if f > 0.0
         g! = create_nonlinear_system_of_equations(material)
-        x0 = [tovoigt(stress); R; tovoigt(X1); tovoigt(X2)]
-        F = similar(x0)
-        res = nlsolve(g!, x0; autodiff = :forward)
+        x0 = state_to_vector(stress, R, X1, X2)
+        res = nlsolve(g!, x0; method=mat.options.nlsolve_method, autodiff=:forward)  # user manual: https://github.com/JuliaNLSolvers/NLsolve.jl
+        converged(res) || error("Nonlinear system of equations did not converge!")
         x = res.zero
-        res.f_converged || error("Nonlinear system of equations did not converge!")
-
-        stress = fromvoigt(SymmetricTensor{2,3,Float64}, @view x[1:6])
-        R = x[7]
-        X1 = fromvoigt(SymmetricTensor{2,3,Float64}, @view x[8:13])
-        X2 = fromvoigt(SymmetricTensor{2,3,Float64}, @view x[14:19])
-        seff = stress - X1 - X2
-        seff_dev = dev(seff)
+        stress, R, X1, X2 = state_from_vector(x)
+
+        # using the new problem state
+        seff_dev = dev(stress - X1 - X2)
         f = sqrt(1.5)*norm(seff_dev) - (R0 + R)
-        dotp = ((f >= 0.0 ? f : 0.0)/Kn)^nn
-        dp = dotp*dtime
-        n = sqrt(1.5)*seff_dev/norm(seff_dev)
+
+        dotp = ((f >= 0.0 ? f : 0.0)/Kn)^nn  # plasticity multiplier, see equations (3) and (4) in Chaboche 2013
+        dp = dotp*dtime  # |dε_p|, using backward Euler (dotp is ∂ε_p/∂t at the end of the timestep)
+        n = sqrt(1.5)*seff_dev/norm(seff_dev)  # Chaboche: a (tensorial) unit direction, s.t. 2/3 * (n : n) = 1
```

Comment:
Mathematically speaking, this is ∂f/∂σ if you wish to add that to the comment :-)

Reply:
`n = ∂f/∂σ`? Sure, I'll add it.